Rewrite it in rust (regex, utils) (#220)
y-scope/log-surgeon: rust/docs/parsing-spec-file.md

## Parsing Specification

> A parsing specification is a list of rules used to determine non-static text in logs.

Reviewer: ```suggestion
A parsing specification is a list of regex rules used to determine semantically meaningful text in logs.
```

> A rule is a `name: "pattern"` pair;

Reviewer: ```suggestion
A rule is a `name: "regex-pattern"` pair;
```

> Log Surgeon supports most common regex syntax/semantics (the outside double-quotes are not part of the pattern).

Reviewer: Link to the regex section here?

> For example, a pattern `[0-9]+` may be used to identify numbers.

Reviewer: ```suggestion
For example, a rule `int: "[0-9]+"` may be used to identify numbers.
```

> In a Parsing Specification File,

Reviewer: What are we following for line length, I assume these can be on 1 line? Just making sure I am not missing some sort of rule.

Author: I've been using 120, though before final release I think have to change to 100 if we're following the rest of clp conventions. But for markdown/text, I also try to break on punctuation/semantically, and prefer more frequent, so that diffs are more fine-grained and relevant to the changed text. Of course, human language sentence structure is complicated and it's subjective whether two fragments that fit on one line should go together (depending on how related they are, e.g. plausibly line 4 above) or not (I'd also be ok breaking up line 4/saying we should "maximize" line breaks)

> rules are listed in priority-order.

Reviewer: ```suggestion
In a Parsing Specification,
rules are listed in order from highest to lower priority.
Parsing first considers longest match, and priority is used as a tie-breaker.
```

These rules are also called root rules;
"regex captures" in a pattern define subrules as `(?<sub_rule_name>sub_rule_pattern)`, arbitrarily nested.
Log Surgeon extracts both root rules and subrules when parsing;
the extracted values are called matches.

Note:
We generally avoid "capture" terminology to avoid ambiguity between the root rule match and sub-match extraction,
which require different implementations internally.
From the outside, Log Surgeon matches and stores both root rule matches and subrule regex captures alike,
so we simply refer to them as "rules", "patterns", and "matches".

A (sub)rule with no child subrules is called a leaf rule;
a root rule whose pattern has no regex captures is also a leaf rule.
For example, the root rule `email: "(?<user>\w+)@(?<hostname>((?<subdomain>\w+)\.)*(?<domain>\w+)\.(?<tld>\w+))"`

> has leaf rules `user`, `subdomain`, `domain`, and `tld`.

Reviewer: Possibly nice to get the user used to seeing the fully qualified names used asap, as this tends to be what shows up in usage anyway (e.g., CLP+ output).
```suggestion
has leaf rules `email.user`, `email.hostname.subdomain`, `email.hostname.domain`, and `email.hostname.tld`.
```

Author: ~~I added/made the change in the next paragraph, since it's where fully qualified names are introduced~~ (edit: saw your next comment)

The fully qualified name of a rule starts with its root rule name and is followed by any/all subrule names,
separated by periods, e.g. `foo.bar.baz`.

> A name denotes the "type" of the matched text,

Reviewer: Suggested moving the qualified name stuff earlier so we can use it in the example.
```suggestion
A name denotes the "type" of the matched text,
```

and the same name/type may have multiple rules/patterns.

> A parsing specification also contains a set of delimiter characters,

Reviewer: Are delimiters mandatory or optional? Also if they are optional, what happens if anchors are used?

> which are additionally used to differentiate between static and non-static text.

Reviewer: Considering the new work on schema-gen, I am starting to dislike the term static-text, as it implies the objective of the schema is to capture variability, when that is more of a heuristic, and the real objective is to capture what the user needs based on their utility function (e.g., some weighted combination of compression ratio and searchability)
```suggestion
which are additionally used to identify matches.
```

### File Format
Again, a parsing specification file primarily consists of a priority-ordered list `name: "pattern"` rules,
plus a delimiter set of characters.

> For example:

Reviewer: ```suggestion
### Example Parsing Specification File
```

The pattern of a placeholder is substituted in-place as a single subexpression,
and on its own **is not** a subrule,
though any subrules in the placeholder's pattern are likewise included in the substitution.
In the following example, `rule1` and `rule2` are semantically equivalent to each other, but not to `rule3` or `rule4`:

```
!greeting: "hello"

rule1: "(?<greeting>)+"
rule2: "(hello)+"
rule3: "hello+"
rule4: "(?<greeting>hello)+"
```

#### Other Notes
- Rules may not match the empty string;
	e.g. the root rule pattern `[0-9]*` or subrule pattern `(?<number>[0-9]*)` is not allowed.
- Placeholders must be defined before they are used.
- Rules must be defined on a single line (they cannot span multiple lines).
- Leading and trailing whitespace on a line are ignored.
- Empty lines are ignored.
- A line starting with a hashtag `#` (ignoring whitespace) is a comment and ignored.
- While a double quote `"` is not special in regex syntax,
	when writing a pattern in a parsing specification file, they must be escaped as `\"`.

### Regex Pattern Syntax
Regexes are ("regular") expressions composed of terms and operators.

Terms are:

> - individual characters/character sets, e.g. `.` (any single character), `a`, `[a-z]`, or `[a-z0-9ABC]`

Reviewer: This includes `\d` etc.?

- parenthesized expressions/subrules, e.g. `(hello)` or `(?<greeting>hello world)`

Operators are, from highest to lowest precedence (and always left-associative):

- postfix repetition, e.g. `a*` (0 or more), `a+` (1 or more), `a?` (0 or 1),
	`a{n}` (exactly `n`), and `a{min,max}` (at least `min`, up to and including `max` times)
- binary concatenation, e.g. `a*b`, equivalent to `(a*)b`
- binary alternation ("or"), e.g. `a*b|c`, equivalent to `((a*)b)|c`

Furthermore, a root rule's pattern may be anchored with a leading `^` or ending `$`,
meaning the match must be preceded/followed by one of the delimiter characters

> (or the very start or end of input).

Reviewer: Make a distinction from what people normally use these symbols for?
```suggestion
(or the very start or end of input), rather than simply the beginning or end of file.
```

which require different implementations internally.
From the outside, Log Surgeon matches and stores both root rule matches and subrule regex captures alike,
so we simply refer to them as "rules", "patterns", and "matches".

A notable distinction is that subrules have a fully qualified name containing its parent/ancestor rules' names;
since a root rule has no parents, its fully qualified name is just its own name.
The fully qualified name of a rule starts with its root rule name and is followed by any/all subrule names,
separated by periods, e.g. `foo.bar.baz`.

A (sub)rule with no child subrules is called a leaf rule;
a root rule whose pattern has no regex captures is also a leaf rule.
For example, the root rule `email: "(?<user>\w+)@(?<hostname>((?<subdomain>\w+)\.)*(?<domain>\w+)\.(?<tld>\w+))"`
has leaf rules `email.user`, `email.subdomain`, `email.domain`, and `email.tld`.

A name denotes the "type" of the matched text,
and the same name/type may have multiple rules/patterns.
For example:

```
timestamp: "\d{2}-\d{2}-\d{4} \d{2}:\d{2}:\d{2}(\.\d{3})?"
timestamp: "\d{2}\\[A-Z][a-z]{2}\\\d{4}:\d{2}\d{2}:\d{2}"
```

> A parsing specification also contains a set of delimiter characters,

Reviewer: We should state somewhere newline is always part of this set.
